drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,305 @@
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 1);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_STATUS_1);
+        self.wait_fpga_busy();
+        self.read_reg(reg::RX_FIFO).to_le_bytes()[0]
+    }
+
+    fn read_flash_status_3(&self) -> u8 {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 1);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_STATUS_3);
+        self.wait_fpga_busy();
+        self.read_reg(reg::RX_FIFO).to_le_bytes()[0]
+    }
+
+    fn flash_write_enable(&self) {
+        self.write_reg(reg::DATA_BYTES, 0);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::WRITE_ENABLE);
+        self.wait_fpga_busy();
+    }
+
+    fn clear_fifos(&self) {
+        self.write_reg(reg::SPICR, 0x8080);
+    }
+
+    fn flash_sector_erase(&mut self, addr: u32) {
+        self.flash_write_enable();
+        self.write_reg(reg::DATA_BYTES, 0);
+        self.write_reg(reg::ADDR, addr);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::BLOCK_ERASE_64KB_4B);
+
+        // Wait for the busy flag to be unset
+        self.wait_flash_busy(Trace::SectorEraseBusy);
+    }
+
+    fn flash_read<'a, B: idol_runtime::BufWriter<'a>>(
+        &mut self,
+        offset: u32,
+        mut dest: B,
+    ) -> Result<(), ()> {
+        let len = dest.remaining_size();
+        assert!(len <= 256);
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, len as u32);
+        self.write_reg(reg::ADDR, offset);
+        self.write_reg(reg::DUMMY_CYCLES, 8);
+        self.write_reg(reg::INSTR, instr::FAST_READ_QUAD_OUTPUT_4B);
+        for i in 0..len.div_ceil(4) {
+            self.wait_fpga_rx();
+            let v = self.read_reg(reg::RX_FIFO);
+            let v = v.to_le_bytes();
+            for (j, byte) in v.iter().enumerate() {
+                let k = i * 4 + j;
+                if k < len {
+                    dest.write(*byte)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn flash_page_program<'a, B: idol_runtime::BufReader<'a>>(
+        &mut self,
+        addr: u32,
+        mut data: B,
+    ) -> Result<(), ()> {
+        let len = data.remaining_size();
+        assert!(len <= 256);
+
+        self.flash_write_enable();
+        self.write_reg(reg::DATA_BYTES, len as u32);
+        self.write_reg(reg::ADDR, addr);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        for i in 0..len.div_ceil(4) {
+            let mut v = [0u8; 4];
+            for (j, byte) in v.iter_mut().enumerate() {
+                let k = i * 4 + j;
+                if k < len {
+                    let Some(d) = data.read() else {
+                        return Err(());
```

Comment:
Should this use an error type? I'm surprised Clippy isn't mad about this tbh.

Reply:
Mildly off-topic, but I've had good(?) use of `Result<T, ()>` in `TryFrom<ParentEnum> for SubEnum` places. In that case the error case is just "it wasn't in the subgroup you checked for". So I'm not _that_ surprised clippy doesn't complain about this: There are some minor use cases where it's not an entirely bad choice.